rfcs/0011-ban-scoring.md

```diff
@@ -0,0 +1,145 @@
+lesser amount. For instance, and invalid hash such that the proof of
+work fails, increments DoS by 50. A previous-block-not-found rejection
+increments DoS by 10. In several cases, transactions are rejected, but
+the DoS is not incremented, such as the "mempool-full" condition.
+
+# Coda at the moment
+
+Bitcoin is a mature codebase, so there are many places where ban scoring has been 
+used. Nonetheless, Bitcoin uses a relatively coarse ban scoring system; only 
+a few ban score increment values are used. In Coda, we could reify scores into a 
+datatype:
+
+```ocaml
+  type ban_score = 
+    | Ban_score_severe
+	| Ban_score_moderate
+	| Ban_score_trivial
+```
+A slightly finer gradation could be used, if desired. The blacklisting system 
+could translate these constructors into numerical scores. Let's call these
+constructors SEV, MOD, and TRV.
+
+In a number of places, the Coda codebase has comments indicating that 
+a peer should be punished, either via a `TODO` or call to `Logger.faulty_peer`.
+Let's classify those places where punishment has been flagged, and annotate 
+them with suggested constructors:
+
+- in `bootstrap_controller.ml`, for bad proofs (SEV), and a validation error when 
+    building a breadcrumb (SEV)
+- in `coda_networking.ml`, when an invalid staged ledger hash is received (SEV), or
+    when a transition sender does not return ancestors (MOD)
+- in `ledger_catchup.ml`, when a root hash can't be found (SEV), or a peer returns an empty list
+    of transitions (instead of `None`) (TRV)
+- in `linked_tree.ml`, for peers requesting nonexistent ancestor paths (MOD)
+- in `parallel_scan.ml`, in `update_new_job` for unneeded merges (?) (MOD)
+- in `staged_ledger.ml`, when a bad signature is encountered when applying a pre-diff (SEV)
+- in `syncable_ledger.ml`, in `num_accounts`, when a content hash doesn't match a stored root hash (SEV), 
+    and in `main_loop`, when a child hash can't be added (MOD)
+- in `catchup_scheduler.ml` and `processor.ml`, when a breadcrumb can't be built from a 
+    transition (SEV) (same failure as in `bootstrap_controller.ml`, above)
+- in `ledger_catchup.ml`, a transition could not be validated (SEV)
+- in `transaction_pool.ml`, a payment check fails (SEV)
+
+At these points in the code, the blacklist API would be called with these constructors.
+
+## Drawbacks
+
+A banning system in necessary to preserve the integrity of the
+network. The only drawback would be if the system is ineffective.
+
+In this RFC, we don't consider how to make the ban scoring persistent,
+which will be required, since nodes stop and re-start.
+
+## Rationale and alternatives
+
+The locations in the code mentioned above are only those already
+flagged with `TODO` or `faulty_peer`. There may be other code
+locations where punishment is warranted.
+
+In the current code, there are often calls to the logger where
+punishment is mentioned in a `TODO`.  There could be an API that calls
```

Comment:
In the trust scores code I have:
```ocaml
  val record : t -> peer -> action -> unit
  (** Record an action a peer took. This may result in a ban event being
      emitted *)
```
Where `action` is a type we functor over. The intent is that calling that function emits a log message, updates the trust score, and potentially writes a ban event to a pipe. The module passed to the functor needs to provide a function `action -> string` so we can write useful log messages. It's an open question whether `action` should be a variant that enumerates everything that affects trust scores or just something like `float * string`. The former keeps all the trust increments together and makes changing them as a set easier, but is likely to lead to a really big variant. I'm kinda leaning towards the latter now, but :man_shrugging:.

Reply:
I like the idea of a variant so we can keep infraction types next to each other and indexed in one spot

Reply:
In the Bitcoin sources, where misbehavior is flagged, sometimes they provide a message, sometimes not. We should always provide a message.